pipelined/bevy_render2/src/primitives/mod.rs

```diff
@@ -0,0 +1,140 @@
+use bevy_ecs::reflect::ReflectComponent;
+use bevy_math::{Mat4, Vec3, Vec3A, Vec4};
+use bevy_reflect::Reflect;
+
+/// An Axis-Aligned Bounding Box
+#[derive(Clone, Debug, Default, Reflect)]
+#[reflect(Component)]
+pub struct Aabb {
+    pub center: Vec3,
+    pub half_extents: Vec3,
+}
+
+impl Aabb {
+    pub fn from_min_max(minimum: Vec3, maximum: Vec3) -> Self {
+        let center = 0.5 * (maximum + minimum);
+        let half_extents = 0.5 * (maximum - minimum);
+        Self {
+            center,
+            half_extents,
+        }
+    }
+
+    /// Calculate the relative radius of the AABB with respect to a plane
+    pub fn relative_radius(&self, p_normal: &Vec3A, axes: &[Vec3A]) -> f32 {
+        // NOTE: dot products on Vec3A use SIMD and even with the overhead of conversion are net faster than Vec3
+        let half_extents = Vec3A::from(self.half_extents);
+        Vec3A::new(
+            p_normal.dot(axes[0]),
+            p_normal.dot(axes[1]),
+            p_normal.dot(axes[2]),
+        )
+        .abs()
+        .dot(half_extents)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Sphere {
+    pub center: Vec3,
+    pub radius: f32,
+}
+
+impl Sphere {
+    pub fn intersects_obb(&self, aabb: &Aabb, model_to_world: &Mat4) -> bool {
+        let aabb_center_world = *model_to_world * aabb.center.extend(1.0);
+        let axes = [
+            Vec3A::from(model_to_world.x_axis),
+            Vec3A::from(model_to_world.y_axis),
+            Vec3A::from(model_to_world.z_axis),
+        ];
+        let v = Vec3A::from(aabb_center_world) - Vec3A::from(self.center);
+        let d = v.length();
+        let relative_radius = aabb.relative_radius(&(v / d), &axes);
+        d < self.radius + relative_radius
+    }
+}
+
+/// A plane defined by a normal and distance value along the normal
+/// Any point p is in the plane if n.p = d
+/// For planes defining half-spaces such as for frusta, if n.p > d then p is on the positive side of the plane.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Plane {
+    pub normal_d: Vec4,
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Frustum {
+    pub planes: [Plane; 6],
+}
+
+impl Frustum {
+    // NOTE: This approach of extracting the frustum planes from the view
+    //       projection matrix is from Foundations of Game Engine Development 2
```

Comment:
small style consistency thing: we don't generally indent in cases like this.

Reply:
Done for any that I touched in this PR.